lang/cs/Org.Apache.REEF.Common/Runtime/Evaluator/HeartBeatManager.cs

```diff
@@ -61,46 +61,50 @@ public class HeartBeatManager : IObserver<Alarm>
 
         private IDriverConnection _driverConnection;
 
-        private EvaluatorSettings _evaluatorSettings;
+        private readonly EvaluatorSettings _evaluatorSettings;
 
         // the queue can only contains the following:
         // 1. all failed heartbeats (regular and event-based) before entering RECOVERY state
         // 2. event-based heartbeats generated in RECOVERY state (since there will be no attempt to send regular heartbeat)
         private readonly Queue<EvaluatorHeartbeatProto> _queuedHeartbeats = new Queue<EvaluatorHeartbeatProto>();
 
-        public HeartBeatManager(EvaluatorSettings settings, IRemoteIdentifier remoteId)
+        [Inject]
+        private HeartBeatManager(
+            EvaluatorSettings settings,
+            [Parameter(typeof(ErrorHandlerRid))] string errorHandlerRid)
         {
-            using (LOGGER.LogFunction("HeartBeatManager::HeartBeatManager"))
+            using (Logger.LogFunction("HeartBeatManager::HeartBeatManager"))
             {
+                _evaluatorSettings = settings;
                 _remoteManager = settings.RemoteManager;
-                _remoteId = remoteId;
-                _evaluatorId = settings.EvalutorId;
+                _remoteId = new SocketRemoteIdentifier(NetUtilities.ParseIpEndpoint(errorHandlerRid));
                 _observer = _remoteManager.GetRemoteObserver(new RemoteEventEndPoint<REEFMessage>(_remoteId));
                 _clock = settings.RuntimeClock;
                 _heartBeatPeriodInMillSeconds = settings.HeartBeatPeriodInMs;
-                _maxHeartbeatRetries = settings.MaxHeartbeatFailures;
-                EvaluatorSettings = settings;
+                _maxHeartbeatRetries = settings.MaxHeartbeatRetries;
+                _evaluatorSettings = settings;
                 MachineStatus.ToString(); // kick start the CPU perf counter
             }
         }
 
-        [SuppressMessage("StyleCop.CSharp.NamingRules", "SA1307:AccessibleFieldsMustBeginWithUpperCaseLetter", Justification = "Intended to be private, exposed now before using future injection")]
-        public EvaluatorRuntime _evaluatorRuntime { get; set; }
+        /// <summary>
+        /// Return EvaluatorRuntime referenced from HeartBeatManager
+        /// Notes: As multiple tangling of the references, it cannot be injected in the constructor even using FutureInjection at this time
+        /// </summary>
+        public EvaluatorRuntime EvaluatorRuntime { get; set; }
```

Comment:
It is set in Evaluator after Evaluator gets EvaluatorRuntime. 

We could use inject future. However, we have triangle reference among EvaluatorRuntime, HeartBeatManager and ContextManager. I tried future injection for different combinations, it doesn't work.